### PR TITLE
dump and load most of couch data

### PR DESCRIPTION
**Pull request by snopoke**
@emord @czue 
buddy @gcapalbo 

* dump data to GZ files
* bundle sql, couch + other data into ZIP
* dump most of couch models including Domain and Toggles

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 13 errors:

* corehq/apps/dump_reload/couch/load.py, line 21 - E501 line too long (117 > 115 characters)
* corehq/apps/dump_reload/management/commands/dump_domain_data.py, line 2 - F401 'json' imported but unused
* corehq/apps/dump_reload/management/commands/dump_domain_data.py, line 8 - F401 'ResourceNotFound' imported but unused
* corehq/apps/dump_reload/management/commands/dump_domain_data.py, line 11 - F401 'Domain' imported but unused
* corehq/apps/dump_reload/management/commands/dump_domain_data.py, line 27 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/dump_reload/management/commands/load_domain_data.py, line 3 - F401 'json' imported but unused
* corehq/apps/dump_reload/management/commands/load_domain_data.py, line 9 - F401 'ResourceNotFound' imported but unused
* corehq/apps/dump_reload/management/commands/load_domain_data.py, line 12 - F401 'Domain' imported but unused
* corehq/apps/dump_reload/management/commands/load_domain_data.py, line 24 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/dump_reload/sql/load.py, line 76 - E501 line too long (117 > 115 characters)
* corehq/apps/dump_reload/tests/test_couch_dump_load.py, line 13 - F401 'FakeCouchDb' imported but unused
* corehq/apps/dump_reload/tests/test_couch_dump_load.py, line 144 - F401 'CommCareMultimedia' imported but unused
* corehq/apps/dump_reload/tests/test_couch_dump_load.py, line 145 - E501 line too long (116 > 115 characters)

File: corehq/apps/dump_reload/sql/load.py [hunk 1]
```diff
@@ -31,60 +33,60 @@
...
+        if chunk:
+            _process_chunk(chunk)
 
-    _reset_sequences(load_stats_by_db.values())
+        _reset_sequences(load_stats_by_db.values())
 
-    loaded_object_count = sum(stat.loaded_object_count for stat in load_stats_by_db.values())
+        loaded_model_counts = Counter()
+        for db_stats in load_stats_by_db.values():
+            model_labels = ('(sql) {}'.format(get_model_label(model)) for model in db_stats.model_counter.elements())
```
Comment (dimagimon) [inline corehq/apps/dump_reload/sql/load.py]: E501 line too long (117 > 115 characters)

File: corehq/apps/dump_reload/tests/test_couch_dump_load.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+            set_default_consumption_for_product(self.domain_name, 'p1', 42),
+            set_default_consumption_for_product(self.domain_name, 'p2', 23),
+            set_default_consumption_for_supply_point(self.domain_name, 'p1', 'clinic1', 80)
+        ]
+
+        self._dump_and_load(objects)
+
+    def test_multimedia(self):
+        from corehq.apps.hqmedia.models import CommCareAudio, CommCareImage, CommCareVideo
+        image_path = os.path.join('corehq', 'apps', 'hqwebapp', 'static', 'hqwebapp', 'img', 'commcare-hq-logo.png')
```
Comment (dimagimon) [inline corehq/apps/dump_reload/tests/test_couch_dump_load.py]: E501 line too long (116 > 115 characters)

File: corehq/apps/dump_reload/tests/test_couch_dump_load.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
+import json
+import os
+import random
+import uuid
+from StringIO import StringIO
+from collections import Counter
+from datetime import datetime, time
```
Comment (dimagimon) [inline corehq/apps/dump_reload/tests/test_couch_dump_load.py]: F401 'datetime' imported but unused
F401 'time' imported but unused

File: corehq/apps/dump_reload/tests/test_couch_dump_load.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        not_expected_docs = []
+        for doc_type in doc_types:
+            expected_docs.append(_make_doc(doc_type, self.domain_name))
+            not_expected_docs.append(_make_doc(doc_type, 'other-domain'))
+
+        self._dump_and_load(expected_docs, not_expected_docs)
+
+    def test_multimedia(self):
+        from corehq.apps.hqmedia.models import CommCareAudio, CommCareImage, CommCareVideo
+        image_path = os.path.join('corehq', 'apps', 'hqwebapp', 'static', 'hqwebapp', 'img', 'commcare-hq-logo.png')
```
Comment (dimagimon) [inline corehq/apps/dump_reload/tests/test_couch_dump_load.py]: E501 line too long (116 > 115 characters)

**Comment by snopoke** [comment]
:white_check_mark:

File: corehq/apps/dump_reload/couch/dump.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    DocTypeIDProvider(['ExportInstance']),
+    DocTypeIDProvider(['ExportDataSchema']),
+    DocTypeIDProvider(['CaseExportInstance']),
+    DocTypeIDProvider(['CaseExportDataSchema']),
+    DocTypeIDProvider(['CustomDataFieldsDefinition']),
+    DocTypeIDProvider(['FixtureOwnership']),
+    DocTypeIDProvider(['FixtureDataType']),
+    DocTypeIDProvider(['FixtureDataItem']),
+    # 'SimplifiedSyncLog'
+    # 'SyncLog' ? don't think new domains use this at all
```
Comment (czue) [inline corehq/apps/dump_reload/couch/dump.py]: this should be true as far as i know

File: corehq/apps/dump_reload/couch/load.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            self._dbs[doc_type] = db
+        return self._dbs[doc_type]
+
+    def load_objects(self, object_strings, force=False):
+        total_object_count = 0
+        for obj_string in object_strings:
+            total_object_count += 1
+            doc = json.loads(obj_string)
+            db = self._get_db_for_doc_type(doc['doc_type'])
+            db.save(doc)
```
Comment (czue) [inline corehq/apps/dump_reload/couch/load.py]: do we not need to do something with `_rev` here (or in the dump command)? I'm surprised couch doesn't complain about this if `_rev` is set.

**Review by czue — APPROVED** [review]
I didn't review everything super closely though it looks good and I think probably not critical to pore over. 

Just to confirm - this is not at all idempotent right? We would have to blow away the DB and start over to run it multiple times?

**Comment by snopoke** [comment]
@czue `_rev` isn't an issue since we're doing a bulk save with
`new_edits=False` which is the replication mode that keeps the rev the same.

It is actually idempotent provided that no changes to data have happened
since the last run (SQL will overwrite with data in the dump and couch will
keep latest version of docs)

**Comment by czue** [comment]
Ah that's quite cool, didn't know about that param.

**Pull request merged**